File: model-library-server-java/src/main/java/iso/sim/server/RunService.java
package iso.sim.server.service;

import iso.sim.server.dto.run.KafkaConfigurationDto;
import iso.sim.server.dto.run.RunStatusResponse;
import iso.sim.server.dto.run.StartModelRunRequest;
import iso.sim.server.dto.run.StartModelRunResponse;
import iso.sim.server.executor.RunExecutionContext;
import iso.sim.server.executor.RunExecutor;
import iso.sim.server.executor.StubRunExecutor;
import iso.sim.server.store.InMemoryRunStatusStore;
import iso.sim.server.store.RunStatusStore;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

public class RunService {
    private final ModelCatalogService modelCatalogService;
    private final RunExecutor runExecutor;
    private final RunStatusStore runStatusStore;

    public RunService(ModelCatalogService modelCatalogService) {
        this(modelCatalogService, new StubRunExecutor(), new InMemoryRunStatusStore());
    }

    public RunService(ModelCatalogService modelCatalogService, RunExecutor runExecutor, RunStatusStore runStatusStore) {
        this.modelCatalogService = modelCatalogService;
        this.runExecutor = runExecutor;
        this.runStatusStore = runStatusStore;
    }

    public StartModelRunResponse startRun(String modelId, StartModelRunRequest request) {
        modelCatalogService.getModel(modelId);
        validateRequest(request);

        String runId = request.getRunId() != null && !request.getRunId().isBlank()
            ? request.getRunId()
            : "run-" + UUID.randomUUID();
        String acceptedAt = Instant.now().toString();
        String statusUrl = "/v1/runs/" + runId;
        RunStatusResponse status = new RunStatusResponse(
            runId,
            modelId,
            "accepted",
            acceptedAt,
            "Run request accepted; backend start is stubbed for Phase 1E"
        );
        runStatusStore.save(status);
        runExecutor.start(new RunExecutionContext(runId, modelId, request));

        return new StartModelRunResponse(
            runId,
            modelId,
            status.getStatus(),
            statusUrl,
            acceptedAt,
            status.getMessage()
        );
    }

    public RunStatusResponse getRunStatus(String runId) {
        RunStatusResponse status = runStatusStore.get(runId);
        if (status == null) {
            throw new RunNotFoundException(runId);
        }
        return status;
    }

    public List<RunStatusResponse> listRuns() {
        return runStatusStore.getAll();
    }

    public RunStatusStore getRunStatusStore() {
        return runStatusStore;
    }

    private void validateRequest(StartModelRunRequest request) {
        if (request.getInitializationParameters() == null) {
            throw new InvalidRunRequestException("'initializationParameters' is required");
        }
        KafkaConfigurationDto kafka = request.getKafka();
        if (kafka == null) {
            throw new InvalidRunRequestException("'kafka' is required");
        }
        if (kafka.getBootstrapServers() == null || kafka.getBootstrapServers().isBlank()) {
            throw new InvalidRunRequestException("'kafka.bootstrapServers' is required");
        }
        if (kafka.getTopic() == null || kafka.getTopic().isBlank()) {
            throw new InvalidRunRequestException("'kafka.topic' is required");
        }
    }
}

File: src/main/kotlin/iso/sim/server/RunService.kt
/*
 *
 *  Foundry for Executable Digital Engineering
 *  Copyright (C) 2026 simlytics.cloud LLC
 *  All rights reserved.
 *
 *  This software was developed under U.S. Government contract(s).
 *  Rights in this software are defined by the applicable contract and
 *  associated DFARS clauses. Unless otherwise specified, this software
 *  is intended to be delivered with Government Purpose Rights (GPR).
 *
 *  This repository is private. No rights are granted except as provided by:
 *  (1) applicable U.S. Government contract(s), or
 *  (2) a separate written license agreement with simlytics.cloud LLC.
 *
 *  Authorized licensees may use and modify this software only in accordance
 *  with the terms of such agreement.
 *
 *  As a condition of any license:
 *  - All modifications and derivative works must be provided back to
 *     simlytics.cloud LLC.
 *  - Redistribution, sublicensing, or third-party sharing is prohibited
 *    without prior written authorization.
 *
 *  Contributions are subject to a Contributor Agreement. By contributing,
 *  you grant simlytics.cloud LLC all rights necessary to use, modify, and
 *  redistribute your contributions without restriction, and you retain
 *  no independent distribution rights.
 *
 */

package iso.sim.server

import java.time.Instant
import java.util.UUID

class RunService(
    private val modelCatalogService: ModelCatalogService,
    private val runExecutor: RunExecutor = StubRunExecutor(),
    private val runStatusStore: RunStatusStore = InMemoryRunStatusStore()
) {
    fun startRun(modelId: String, request: StartModelRunRequest): StartModelRunResponse {
        modelCatalogService.getModel(modelId)
        validateRequest(request)

        val runId = request.runId?.takeIf { it.isNotBlank() } ?: "run-${UUID.randomUUID()}"
        val acceptedAt = Instant.now().toString()
        val statusUrl = "/v1/runs/$runId"
        val status = RunStatusResponse(
            runId = runId,
            modelId = modelId,
            status = "accepted",
            acceptedAt = acceptedAt,
            message = "Run request accepted; backend start is stubbed for Phase 1E"
        )
        runStatusStore.save(status)
        runExecutor.start(
            RunExecutionContext(
                runId = runId,
                modelId = modelId,
                request = request
            )
        )

        return StartModelRunResponse(
            runId = runId,
            modelId = modelId,
            status = status.status,
            statusUrl = statusUrl,
            acceptedAt = acceptedAt,
            message = status.message
        )
    }

    fun getRunStatus(runId: String): RunStatusResponse {
        return runStatusStore.get(runId) ?: throw RunNotFoundException(runId)
    }

    fun getRunStatusStore(): RunStatusStore {
        return runStatusStore
    }

    private fun validateRequest(request: StartModelRunRequest) {
        if (request.initializationParameters == null) {
            throw InvalidRunRequestException("'initializationParameters' is required")
        }
        val kafka = request.kafka ?: throw InvalidRunRequestException("'kafka' is required")
        if (kafka.bootstrapServers.isNullOrBlank()) {
            throw InvalidRunRequestException("'kafka.bootstrapServers' is required")
        }
        if (kafka.topic.isNullOrBlank()) {
            throw InvalidRunRequestException("'kafka.topic' is required")
        }
    }
}

class InvalidRunRequestException(message: String) : RuntimeException(message)

class RunNotFoundException(runId: String) : RuntimeException("Run '$runId' was not found")
